featuretools/entityset/entity.py

```diff
@@ -62,61 +62,39 @@ def __init__(self, id, df, entityset, variable_types=None,
+                Otherwise, assume index exists in dataframe.
         """
         assert is_string(id), "Entity id must be a string"
         assert len(df.columns) == len(set(df.columns)), "Duplicate column names"
+        for c in df.columns:
+            if not is_string(c):
+                raise ValueError("All column names must be strings (Column {} "
+                                 "is not a string)".format(c))
+        if time_index is not None and time_index not in df.columns:
+            raise LookupError('Time index not found in dataframe')
+
+        created_index, index, df = create_index(index, make_index, df)
+        if index not in variable_types:
+            variable_types[index] = vtypes.Index
+
         self.data = {"df": df,
                      "last_time_index": last_time_index,
                      }
         self.encoding = encoding
         self._verbose = verbose
         self.created_index = created_index
-        self.convert_all_variable_data(variable_types)
         self.id = id
         self.entityset = entityset
-        variable_types = variable_types or {}
-        self.index = index
-        self.time_index = time_index
-        self.secondary_time_index = secondary_time_index or {}
-        # make sure time index is actually in the columns
-        for ti, cols in self.secondary_time_index.items():
-            if ti not in cols:
-                cols.append(ti)
-
-        relationships = [r for r in entityset.relationships
-                         if r.parent_entity.id == id or
-                         r.child_entity.id == id]
-
-        link_vars = [v.id for rel in relationships for v in [rel.parent_variable, rel.child_variable]
-                     if v.entity.id == self.id]
-
-        inferred_variable_types = self.infer_variable_types(ignore=list(variable_types.keys()),
-                                                            link_vars=link_vars)
-        for var_id, desired_type in variable_types.items():
-            if isinstance(desired_type, tuple):
-                desired_type = desired_type[0]
-            inferred_variable_types.update({var_id: desired_type})
-
-        self.variables = []
-        for v in inferred_variable_types:
-            # TODO document how vtype can be tuple
-            vtype = inferred_variable_types[v]
-            if isinstance(vtype, tuple):
-                # vtype is (ft.Variable, dict_of_kwargs)
-                _v = vtype[0](v, self, **vtype[1])
-            else:
-                _v = inferred_variable_types[v](v, self)
-            self.variables += [_v]
 
-        # do one last conversion of data once we've inferred
-        self.convert_all_variable_data(inferred_variable_types)
+        variable_types = variable_types or {}
+        secondary_time_index = secondary_time_index or {}
+        self.create_variables(variable_types, index, time_index, secondary_time_index)
```

Comment:
i think this should be `self._create_variables` since a user should never need call it, right?